### ISSUE TITLE
Implement MiniRocket Feature Representation Model

### ISSUE THREAD

**Issue by sandrohuni** [hydrosolutions/long-term-forecasting#45]
# Implement MiniRocket Feature Representation Model

## Problem Statement

Currently, our monthly discharge forecasting models rely on hand-crafted features and traditional feature engineering approaches. We want to explore MiniRocket-based unsupervised time series representation learning to potentially improve forecast accuracy and model robustness, particularly for basins with limited data or irregular patterns.

## Objective

Implement a new forecasting model that uses MiniRocket for unsupervised feature representation learning followed by linear/tree-based regression models, while maintaining compatibility with our existing operational pipeline.

### Core Requirements
- Use MiniRocket for time series feature representation
- Proper LOOCV fitting (no data leakage)
- Minimal changes to existing codebase
- Operational compatibility with current deployment pipeline
- Support Ridge, Lasso, XGBoost, LightGBM downstream models

### Architecture Overview
```
Input Data → Existing Preprocessing → MiniRocket Features → Downstream Models → Predictions
```

### Phase 1: Feature Representation Framework
- [ ] **Implement MiniRocketFeatureExtractor**
  - File: `monthly_forecasting/scr/MiniRocketFeatureExtractor.py`
  - Clean interface working with existing preprocessing
  - Proper time series window creation for multi-variate data
  - Transform training/test data consistently
  - Handle multiple basins efficiently

### Phase 2: Model Implementation
- [ ] **Create MiniRocketRegressor**
  - File: `monthly_forecasting/forecast_models/MiniRocketRegressor.py`
  - Inherit from `BaseForecastModel`
  - Use existing preprocessing pipeline unchanged
  - Add MiniRocket representation learning after preprocessing
  - Support Ridge, Lasso, ElasticNet, XGBoost, LightGBM downstream models

- [ ] **Implement proper LOOCV**
  - Fit MiniRocket separately for each LOOCV fold
  - No data leakage between train/test
  - Proper artifact saving/loading
  - Operational deployment compatibility

### Phase 3: Configuration & Integration
- [ ] **Extend configuration system**
  - Add `minirocket_config` section to existing configs
  - MiniRocket-specific parameters (num_kernels, input_length)
  - Downstream model selection
  - Example configurations

- [ ] **Integration with existing pipeline**
  - Compatible with current evaluation scripts
  - Works with existing data loading
  - Maintains operational prediction interface
  - Proper model saving/loading

### Phase 4: Testing & Validation
- [ ] **Unit tests**
  - Test feature extractor functionality
  - Test data format handling
  - Test configuration parsing
  - Test model serialization

- [ ] **Integration tests**
  - End-to-end pipeline testing
  - LOOCV validation correctness
  - Operational prediction workflow
  - Performance benchmarking vs SciRegressor

- [ ] **Functionality tests**
  - Multi-basin prediction accuracy
  - Handle missing data gracefully
  - Memory efficiency testing

## File Structure
```
monthly_forecasting/
├── scr/
│   └── MiniRocketFeatureExtractor.py  # MiniRocket feature extractor
├── forecast_models/
│   └── MiniRocketRegressor.py         # New model implementation
├── tests/
│   ├── unit/
│   │   └── test_minirocket_extractor.py
│   └── functionality/
│       └── test_minirocket_regressor.py
└── example_config/
    └── MINIROCKET_MODEL/
        ├── general_config.json
        ├── model_config.json
        └── feature_config.json
```

## Configuration Example
```json
{
  "model_name": "MiniRocketRegressor",
  "model_type": "minirocket_regressor",
  "minirocket_config": {
    "num_kernels": 1000,
    "input_length": 30,
    "random_state": 42
  },
  "downstream_models": ["ridge", "xgboost", "lgbm"],
  "feature_cols": ["discharge", "P", "T", "SWE"],
  "prediction_horizon": 30,
  "target": "target"
}
```

## Success Criteria
- [ ] Model trains without errors on existing dataset
- [ ] Proper LOOCV validation with no data leakage
- [ ] Performance metrics comparable to existing baselines
- [ ] Operational prediction pipeline works end-to-end
- [ ] Clean integration with existing codebase (no modifications to preprocessing)
- [ ] Comprehensive test coverage (>80%)
- [ ] Documentation and usage examples

## Technical Constraints
- **No modifications** to existing `SciRegressor` or preprocessing pipeline
- **Minimal dependencies**: Use existing scikit-learn MiniRocket implementation
- **Memory efficient**: Handle large multi-basin datasets
- **Backwards compatible**: Existing configurations should continue working
- **Operational ready**: Support real-time prediction deployment

## Research Context
- **MiniRocket**: Already used in existing `SciRocket.py`, extend for better representation learning
- **Hydrological Context**: Monthly discharge prediction with multi-variate inputs (meteorological, snow, glacial data)
- **Focus**: Pure feature representation approach vs existing combined approach

## Expected Timeline
- **Week 1**: MiniRocket feature extractor implementation
- **Week 2**: Model implementation + LOOCV integration  
- **Week 3**: Testing, validation, and documentation
- **Total**: ~3 weeks for complete implementation

## Key Differences from Existing SciRocket
- Focus purely on feature representation (no built-in downstream models)
- Cleaner separation between representation learning and regression
- Better LOOCV implementation with proper data isolation
- More flexible downstream model selection

## Acceptance Criteria
- All tasks marked as completed with ✅
- Model achieves NSE > 0.9 on validation set
- No data leakage in LOOCV validation
- Operational prediction generates results within 5 minutes
- Test coverage > 80%
- Documentation includes usage examples

---

**Priority**: High  
**Complexity**: Medium  
**Estimated Effort**: 3 weeks  
**Dependencies**: None (builds on existing infrastructure)
